### DISCUSSION TITLE
Manuscript - How to automate flow chart creation

### DISCUSSION THREAD

**Discussion by shijiaSMH** [GEMINI-Medicine/Rgemini#91] [Ideas]
## Why automating flow chart creation
Manuscript usually requires flow chart for cohort creation. Usually a separate word doc is used to create publishable flow chart since steps involved are likely differently from what's originally planned in the protocol. 

### Benefits of automation
- Decrease manual burden of formatting 
- Reduce rates of mistakes resulted from manual entries

## Recommended package and its installation
`install.packages('DiagrammeR')`

### Code example
```
# create a dataset with each number corresponding to each cohort creation step
cohort <- list(a=620000, b=550000, c=540000, d=520000)

# draw flow chart
DiagrammeR::grViz("
digraph graph2 {

graph [layout = dot]

# node definitions with substituted label text
node [shape = rectangle, width = 4, fillcolor = Biege]
a [label = '@@1']
b [label = '@@2']
c [label = '@@3']
d [label = '@@4']

a -> b -> c -> d

}

# edit inclusion/exclusion criteria
[1]: paste0('Include patients admitted to or discharged from GIM service (n = ', cohort$a, ')')
[2]: paste0('Exclude patients admitted from any route other than emergency (n = ', cohort$b, ')')
[3]: paste0('Exclude patients with any surgery diagnosis (n = ', cohort$c, ')')
[4]: paste0('Include valid physicians who have treated at least 100 encounters  (n = ', cohort$d, ')')
")

```
 ### Flow chart output
![image](https://github.com/GEMINI-Medicine/Rgemini/assets/105071952/1136c879-3efd-4763-b8ee-cc8fd236c837)

**Comment by shijiaSMH** [comment]
# Ideas for further customization
## Function to be more user friendly
- Current set up is easy for analysts to navigate, but not necessarily for beginners to figure out which codes to change. 
- Propose to develop a function to make users to just input how many steps they need for cohort creation, and enter corresponding text as function indicator. Then output is the flow chart screenshot stated above.
- An imaginary function: flow_chart_creation(4, 'Include patients admitted to or discharged from GIM service', 'Exclude patients admitted from any route other than emergency', 'Exclude patients with any surgery diagnosis', 'Include valid physicians who have treated at least 100 encounters')
- Challenge: DiagrammeR::grViz syntax have cohort$variable inside paste0, so can be hard to realize the proposed functionality.

## Streamline output
- As manuscript requires word doc, it's desirable for flow chart to be in this format as well.
- For proposed function, we can then add a T/F indicator for whether to output word of flow chart.
- On the other hand, may not be necessary, since knitting a markdown can get this done easily

@loffleraSMH thoughts?

**Reply by loffleraSMH** [reply]
Sounds like what you are proposing would be a wrapper function for `DiagrammeR::grViz()` - less for customization, but mostly to facilitate usability of the existing functionality. I think that's fine, feel free to open an issue for this. Let's discuss details for this in the issue so we have a cleaner record of anything that's been discussed regarding the proposed function (and let others chime in too).
